Intrinsics States.JsonMerge

kbrock opened the pull request:
Part of #64

Ruby missing a `Boolean` is frustrating me more and more.

This adds `JsonMerge`.

Please share your thoughts on how to handle the Boolean Attribute.

The other solution I had was a multi argument that handles a crystal style argument and it went through each of them. But it still had more modification in the comparison portion of the check code

Another solution is to just specify Object, and then manually make a check, but this required duplication of the error message.

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -9,6 +9,11 @@ class Workflow
     class IntrinsicFunction
       class Transformer < Parslet::Transform
         OptionalArg = Struct.new(:type)
+        class Boolean
```
Super-mega-minor
```suggestion

        class Boolean
```

Fryguy commented:
Wow...it's that simple!?  LOL I was so scared to do this one haha.

Fryguy commented:
oops...accidental close.

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -9,6 +9,11 @@ class Workflow
       class Transformer < Parslet::Transform
         OptionalArg = Struct.new(:type)
+        class Boolean
+          def self.to_s ; "Boolean" ; end
+          # NOTE: this is backwards
+          def self.kind_of?(value) ; value.kind_of?(TrueClass) || value.kind_of?(FalseClass) ; end
```
I agree with Rubocop - not a fan of single-line methods, even if simple, so let's just blow this out.

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -9,6 +9,11 @@ class Workflow
         OptionalArg = Struct.new(:type)
+        class Boolean
+          def self.to_s ; "Boolean" ; end
+          # NOTE: this is backwards
```
Can you add more detail here?  What makes it backwards in your opinion?  Are there tests demostrating this backwardness?

kbrock replied:
we do `Boolean.kind_of?(obj)` instead of `obj.kind_of?(Boolean)`

That is why we have a weird.

I have renamed to `cover?` and it reads a little better

Fryguy commented on lib/floe/workflow/intrinsic_function/transformer.rb:
```diff
@@ -50,7 +55,8 @@ def check_types(args, function, signature)
               type = type.type if type.kind_of?(OptionalArg)
 
-              raise ArgumentError, "wrong type for argument #{index + 1} to #{function} (given #{arg.class}, expected #{type})" unless arg.kind_of?(type)
+              matched = type == Boolean ? type.kind_of?(arg) : arg.kind_of?(type)
```
I wonder if we can do this better with something like `===`, which can handle kind_of logic?

kbrock replied:
ooh. we are not asking if type is a kind of Boolean, we are asking if type == Boolean.
Introducing `===` makes the tests fail

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,23 @@
+      end
+
+      it "fails with invalid args" do
+        expect { described_class.value("States.JsonMerge($.left, $.right, 5)", ctx, {"left" => {"a" => "la"}, "right" => {"b" => "rb"}}) }.to raise_error(ArgumentError, "wrong type for argument 3 to States.JsonMerge (given Integer, expected Boolean)")
```
Please include failure tests for every parameter (i.e. non-Hash in positions 1 and 2)

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,23 @@
     end
 
+    describe "States.JsonMerge" do
+      let(:ctx) { {"context" => {"baz" => "qux"}} }
```
This context is not needed, so it complicated the test - Can we remove it and just hardcode an empty context like we do for the other tests?

Fryguy commented:
I think the addition of Boolean is great.  Only thought I had was maybe to implement `===`, then we can unify the check by switching from kind_of? to `===`.  Not sure it will work, but there's so many tests, that we'll know pretty quickly.  Even so, I'm happy to merge as is, and this can be a followup.

Fryguy commented:
Overall LGTM :+1: - A few comment on specs, then this is good to go.

Fryguy commented on lib/floe/workflow/intrinsic_function/parser.rb:
```diff
@@ -64,6 +64,7 @@ class Parser < Parslet::Parser
           :states_base64_encode,   "States.Base64Encode",
           :states_base64_decode,   "States.Base64Decode",
           :states_hash,            "States.Hash",
+          :states_json_merge,      "States.JsonMerge",
```
This order is not alphabetical - it currently matches the order as listed in the stateslanguage.net spec.  The same goes for the order of the methods within the transformer.

kbrock commented:
Looks like we need to introduce `deep_merge` -- not sure if we want to bring in a gem or want to implement this. (I was surprised to see that `deep_merge` gem was one of yours. YAY)

The problem with `===` and `kind_of?` is the method is defined on `Object` not `Boolean`. And we will not get called in those instances.

@Fryguy what do you think about defining `Boolean`?
I'm thinking of adding Boolean to other places too.

kbrock commented:
update:
- Rebase
- Introduce `Boolean` to remove special cases around `kind_of?`
- Implement `deep_merge`
- Add more specs.

update:
- Fixed spacing in hash (rubocop)

Fryguy commented:
This need a rebase after #256

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,30 @@
     end
 
+    describe "States.JsonMerge" do
+      it "it merges with right hand precidence" do
```
```suggestion
      it "it merges with right hand precedence" do
```

Fryguy commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -450,6 +450,30 @@
+        expect(result).to eq({"a" => "la", "b" => "rb"})
+      end
+
+      it "it deep merges with right hand precidence" do
```
```suggestion
      it "it deep merges with right hand precedence" do
```

Fryguy commented on lib/floe/boolean.rb:
```diff
@@ -0,0 +1,10 @@
+module Floe
+  module Boolean
```
Not sure I like this here - Originally you had it under Floe::Workflow::IntrinsicFunction::Boolean, which made more sense to keep it contained there.  Also, not sure I like messing with the internals of TrueClass and FalseClass.

Would like another opinion though - @agrare ?

kbrock commented:
Update:
- rebase (resolved conflict)
- fixed spelling
- added number of parameters check

Still unsure what to do about Boolean.
Thinking that a MultiType (ala Crystal `TrueClass|FalseClass` / `[TrueClass, FalseClass]`) may be the best route.

Fryguy commented:
Coincidentally, I added "Union" types to my States.Format PR, because I needed that.  Perhaps you can use that after that one is merged?

kbrock commented:
Update:
- dropped Boolean and used Union type
